oauth/scope.js

```diff
@@ -0,0 +1,514 @@
+
+'use strict';
+
+const { URL } = require('url');
+
+// These character ranges are from the OAuth RFC,
+// https://tools.ietf.org/html/rfc6749#section-3.3
+const VALID_SCOPE_VALUE = /^[\x21\x23-\x5B\x5D-\x7E]+$/;
+
+const VALID_SHORT_NAME_VALUE = /^[a-zA-Z0-9_]+$/;
+const VALID_FRAGMENT_VALUE = /^#[a-zA-Z0-9_]+$/;
+
+
+/**
+ * A `Scope` is a set of validated scope string values, against which
+ * we can perform various membership checks and set-like operations
+ * such as:
+ *
+ *  - Checking whether one set of scope values wholly or partly
+ *    contains another set.
+ *
+ *  - Filtering out scope values that are (or are not) implied
+ *    by another set.
+ *
+ *  - Aggregating values into a single set.
+ *
+ * We expect `Scope` objects will be used to process user-provided
+ * data, so it's important that they provide efficient operations
+ * that do not open potential DoS vectors.  A naive item-by-item
+ * comparison of two sets can easily be accidentally quadratic.
+ *
+ * The key insight required for achieving this is that, for given
+ * a scope string `v`, we can enumerate the set of all possible values
+ * that imply `v`, and the size of this set of linear in the length
+ * of `v`.  These values are called the "implicants" of `v` and
+ * by pre-calculting and storing them, we are able to perform many
+ * `Scope` operations using simple set lookups.
+ *
+ * To that end, `Scope` instances store their data internally as a
+ * pair of cross-referenced maps, `this._values` and `this._implicants`,
+ * both of which are keyed by scope string values.
+ *
+ * The `_values` map has as its keys the actual scope values contained
+ * in the set, and the corresponding items are the set of implicants
+ * for each scope value.  Conversely, the `_implicants` map has as its
+ * keys every possible implicant of a value in the set, and the corresponding
+ * items are the set of scope values that are implied by that implicant.
+ *
+ * Thus:
+ *
+ *  - if `v in this._values` then the value `v` is a member of this
+ *    set, and `this._values[v]` gives all possible values that might
+ *    imply it.
+ *
+ *  - if `v in this._implicants` then the value `v` implies one or more
+ *    value in this set, and `this._implicants[v]` gives all the values
+ *    in this set that are implied by it.
+ *
+ * This scheme allows us to implement efficient membership checks, set
+ * comparison, and set union with deduplication of redundant values.
+ *
+ * The tradeoff here is memory usage, as we have to store the full
+ * set of implicants for each scope value.  We expect that this will not
+ * be too problematic in practice, because the size of this set is linear
+ * in the length of the scope value. A more memory-friendly implementation
+ * could use a trie-like data structure to efficiently store all implicants,
+ * with prefix compression.
+ *
+ *
+ */
+class Scope {
+
+  constructor(values, ...parentScopes) {
```

Comment:
I refactored this away FWIW, all that places that used to reply on this now call `_addScope` directly with the values in question, rather than trying to look up the pre-parsed set of implicants from other objects.